In [1]:
import google.colab
%pip install roboticstoolbox-python>=1.0.2
%pip install colored==1.4.4

  Preparing metadata (setup.py) ... done
  Created wheel for colored: filename=colored-1.4.4-py3-none-any.whl size=14249 sha256=a36a21ff5cd76d9310fe07ede5b58a5214e0a4c53abedd4fb8bf6116c8d8a002
  Stored in directory: /root/.cache/pip/wheels/e7/c3/07/fabb0941ff5df7a487d43a67081273045536cc96d4d0e816b4
Successfully built colored
  Attempting uninstall: colored
    Found existing installation: colored 2.2.3
    Uninstalling colored-2.2.3:
      Successfully uninstalled colored-2.2.3


In [2]:
import numpy as np
import roboticstoolbox as rtb
from spatialmath import *
from math import pi
import matplotlib.pyplot as plt
from matplotlib import cm
np.set_printoptions(linewidth=100, formatter={'float': lambda x: f"{x:8.4g}" if abs(x) > 1e-10 else f"{0:8.4g}"})

%matplotlib notebook

Rozwiązanie zagadnienia kinematyki odwrotnej, podobnie jak zagadnienia kinematyki prostej, trzeba rozpocząć od stworzenia modelu manipulatora.

In [3]:
L1 = rtb.DHLink(d=0.645, alpha=pi/2, theta=0.0, a=0.330,)
L2 = rtb.DHLink(d=0.0, alpha=0.0, theta=0.0, a=1.35)
L3 = rtb.DHLink(d=0.0, alpha=pi/2, theta=0.0, a=0.115)
L4= rtb.DHLink(d=1.420, alpha=-pi/2, theta=0.0, a=0.0)
L5= rtb.DHLink(d=0.0, alpha=pi/2, theta=0.0, a=0.0)
L6= rtb.DHLink(d=0.215, alpha=0.0, theta=0.0, a=0.0)
robot = rtb.DHRobot([L1, L2, L3, L4, L5, L6])

W następnym kroku konieczne jest określenie położenia i orientacji końcówki manipulatora, dla której zagadnienie  ma zostać rozwiązane. Pozycję tą należny przedstawić w postaci macierzy homogenicznej. Jednym z najprostszych sposobów jest określenie macierzy translacji i rotacji. <br>
Macierz translacji stworzymy obliczając ją dla przykładu z kinematyki prostej. W ten sposób manipulator ustawi się ostatecznie w tym samym miejscu.

Rozwiązanie rozwiązania zagadnienia kinematyki odwrotnej można otrzymać wywołując na obiekcie robota metodę **ikine_LM** i podając jako argument macierz jednorodną.
```py
sol = robot.ikine_LM(T)
```
W zwracanym obiekcie *sol* zawarte są informacje o rozwiązaniu problemu, w tym współrzędne złączowe i informacja o tym czy udało się osiągnąć zadaną pozycję.

In [5]:
i = 19
new_q = [pi/125*i, pi/2+(-pi/225*i), -pi/100*i, -pi/50*i, -pi/100*i, -pi/100*i]
T = robot.fkine(new_q)
print(T)
sol = robot.ikine_LM(T)
sol

  -0.3013    0.7579    0.5787    1.63      
   0.7932    0.536    -0.289     0.717     
  -0.5292    0.3719   -0.7626    0.7805    
   0         0         0         1         



IKSolution(q=array([  -2.664,    2.229,    2.972,  -0.8133,   0.8028,    2.054]), success=True, iterations=15, searches=1, residual=1.2619959472657988e-11, reason='Success')

In [ ]:
print(sol.success)
print(sol.q)

True
[  0.4775    1.306  -0.5969   -1.194  -0.5969  -0.5969]


In [ ]:
with open("wynik.txt", "w") as plik:
    plik.write(','.join(map(str,sol.q)))
    plik.write('\n')

First need path planning:
found nothing in google <br>
Need to figure out on my own:
muszę założyć że już znam punkty,
jak auto wjeżdża na podnośnik to zatrzymuje się w konkretnym miejscu.
Wtedy wiem że spaw idzie na wysokości np 1.2 m i idze pod kątem 30
So generate the workspace first
czyli znam te punkty i one będą w postaci macierzy T
to teraz od jednego do drugiego
Joint vs Task się pojawia (oba dochodzą do punktu)
zróbmy Task czyli jak mam dwa punkty to jak wygenerować te pośrednie?
czyli Interpolacja?
spline itp
na wykresie wartosc kazdej zmiennej w T i do tego czas
ale czy ja je moge zmieniac osobno?
https://www.youtube.com/watch?v=Fd7wjZDoh7g
trapezoidcal, polynomial trajectories like cubic sounds good
